### PR TITLE
[unitaryHACK] Add the YY gate

### PR DESCRIPTION
**Pull request by paniash**
Adds IsingYY gate functionality.

**Related GitHub Issues:** #1347

### CODE REVIEW

File: pennylane/ops/qubit.py
Comment (nahumsa) [inline pennylane/ops/qubit.py:1847]:
```suggestion
    * Gradient recipe: 
    
    .. math::
         \frac{d}{d\phi} f(YY(\phi)) 
         = \frac{1}{2} \left[ f(YY(\phi + \pi/2)) - f(YY(\phi - \pi/2)) \right]

     where :math:`f` is an expectation value depending on :math:`YY(\phi)`.
```

Comment (nahumsa) [reply pennylane/ops/qubit.py:1847]: Maybe it would be better to have an indention block for the math equation, just like other `Ising` gates?

I changed `\theta` to `\phi` for consistency.

**Comment by nahumsa** [comment]
I think that it is missing one test for the `isingYY` [here](https://github.com/PennyLaneAI/pennylane/blob/9a8d1b70f69e8407e40291d8a5a12f1fc9842168/tests/ops/test_qubit_ops.py#L1980-L1998). which could just add `"YY"` on the parametrized test :  `@pytest.mark.parametrize("pauli_word", ["XX", "ZZ"])`

**Comment by paniash** [comment]
> I think that it is missing one test for the `isingYY` [here](https://github.com/PennyLaneAI/pennylane/blob/9a8d1b70f69e8407e40291d8a5a12f1fc9842168/tests/ops/test_qubit_ops.py#L1980-L1998). which could just add `"YY"` on the parametrized test : `@pytest.mark.parametrize("pauli_word", ["XX", "ZZ"])`.
> 
> But I'm not sure about it. I was just looking that this was done on the `IsingZZ` PR #1199.

Thanks for pointing out! Also, the only pending changes looking at how similar PRs for ZZ and XX were done, is in ` tests/devices/test_default_qubit.py`. Will do that and make it ready for review soon enough.

**Comment by paniash** [comment]
I've implemented everything similar to XX and ZZ (as per my knowledge). Let me know if it's ready to be reviewed.

**Comment by nahumsa** [comment]
> I've implemented everything similar to XX and ZZ (as per my knowledge). Let me know if it's ready to be reviewed.

LGTM. I think it's ready for a review! Did you check if the tests are passing?

**Comment by paniash** [comment]
@nahumsa I haven't run tests on my local machine (some build errors come up). Shall I open for review and allow github actions to run the tests?

**Comment by josh146** [comment]
Nice work! However, it looks like two of the CI checks are failing — simply leave a comment here once you have them passing, or if you’re not sure why the CI is failing, feel free as well to ask and we can look into it 🙂

**Comment by paniash** [comment]
@josh146 Thanks! I'm having some build errors come up when running `make test` so probably that's the reason why the `core-tests` test is failing. Regarding the `formatting check`, could you let me know how I can fix it? Thanks! :-)

**Comment by josh146** [comment]
> Regarding the formatting check, could you let me know how I can fix it?

Sure! This just means that you need to run the `black` code formatter, with line length 100, on both the `pennylane` and `tests` folder _inside_ the repo:

```
pip install black --upgrade
black -l 100 pennylane tests
```

**Comment by josh146** [comment]
@paniash almost there, looks like just one test is failing:

```python
 =================================== FAILURES ===================================
__________________ TestOperations.test_isingyy_decomposition ___________________
Traceback (most recent call last):
  File "/home/runner/work/pennylane/pennylane/tests/ops/test_qubit_ops.py", line 740, in test_isingyy_decomposition
    assert res[1].wires == Wires([2])
AssertionError: assert <Wires = [3]> == <Wires = [2]>
  At index 0 diff: 3 != 2
  Use -v to get the full diff
```

File: tests/devices/test_default_qubit.py
Comment (ixfoduap) [inline tests/devices/test_default_qubit.py:1]: Are there any changes actually done here other than formatting? I couldn't find any

File: tests/ops/test_qubit_ops.py
Comment (ixfoduap) [inline tests/ops/test_qubit_ops.py:733]: You should probably test over more parameters, for example negative ones

File: tests/ops/test_qubit_ops.py
Comment (ixfoduap) [inline tests/ops/test_qubit_ops.py:757]: Awesome!

File: tests/ops/test_qubit_ops.py
Comment (ixfoduap) [inline tests/ops/test_qubit_ops.py:758]: I don't see any other tests of the operation?

Most importantly, we should be testing that gradients and adjoints are computed correctly!

File: tests/devices/test_default_qubit.py
Comment (paniash) [reply tests/devices/test_default_qubit.py:1]: Ah, I just changed 2020 to 2021.

File: .github/CHANGELOG.md
Comment (antalszava) [inline .github/CHANGELOG.md:385]: Hi @paniash, it seems that new additions to the `master` branch are causing a conflict here (and as a result some contributors are missing from the PR). Could you perhaps change the list locally to the one that is there in `master` and then add your name where it'd fit in alphabetical order? :slightly_smiling_face:

Comment (paniash) [reply .github/CHANGELOG.md:385]: Sorry about that! It should be fixed now. :-)

Comment (paniash) [reply .github/CHANGELOG.md:385]: Looks like there is some additional merge conflict that arises. Sorry about this. I'll fix it as soon as I can.

Comment (antalszava) [reply .github/CHANGELOG.md:385]: That's okay, no worries! Locally, once thing that can be done is to pull `master`, switch to the dedicated branch and then go for `git merge master`. This will merge the changes from master and show if there are conflicts for each file.

Then going through each file with conflicts and resolving them helps. `git` will put `HEAD` and `master` markers (the branch name with which there's a conflict) in the file with visual lines too. Resolving the conflict goes as comparing the version of the source file from the pull request with the version of the source file on `master`. Some conflicts can be non-trivial, let us know if something like this comes up.

Comment (paniash) [reply .github/CHANGELOG.md]: @antalszava Sorry about the delay, was caught up with some personal stuff. I've fixed the merge conflict. Let me know if it's alright. :-)

**Comment by paniash** [comment]
@antalszava @josh146 Looks like there is a failing test with the `test_isingyy_decomposition()` function. Not sure what the issue is, and would be glad if you could help. :-)

**Comment by josh146** [comment]
Hi @paniash! It seems that the test is failing because the decomposition does not match the Ising matrix.

Have you double checked that the decomposition is correct?

If so, it might be worth double checking that the test is performing the correct matrix multiplication!

**Comment by crazy4pi314** [comment]
Hey @paniash  and the PennyLane team! I wanted to let folks know I am extending the unitaryHACK deadline till the end of this week for PRs that were already in progress to give some time to iterate with maintainers. Let me know if you have questions, or if I can be of help in getting this issue closed!

**Comment by paniash** [comment]
Looks like there is a coverage failure in `pennylane/ops/qubit.py`. I'd be glad if someone could help me rectify this.

File: pennylane/ops/qubit.py
Comment (antalszava) [inline pennylane/ops/qubit.py:1843]: How come these parts, (unrelated to IsingYY) of the file were changed?

Comment (paniash) [reply pennylane/ops/qubit.py:1843]: To maintain consistency in variable names, I changed them.

**Comment by antalszava** [comment]
Hi @paniash, there seems to be a warning related to complex code in that file. Noticed that there were some changes made to some parts of the file other than the IsingYY addition. Maybe reverting those will help.

File: pennylane/devices/tests/test_gates.py
Comment (antalszava) [inline pennylane/devices/tests/test_gates.py:199]: Hi @paniash, how come such formatting changes were applied? We'd usually be formatting files with `black -l 100 pennylane` where[ `pennylane` is the subdirectory in the repo](https://github.com/PennyLaneAI/pennylane/tree/master/pennylane). Same for the `tests` subdirectory.

Comment (paniash) [reply pennylane/devices/tests/test_gates.py:199]: @antalszava I applied the formatting as suggested. Not sure what happened. However, I ran it once again on the subdirectory `pennylane` and `tests` and got the following output:

```
$ black -l 100 tests pennylane
All done! ✨ 🍰 ✨
282 files left unchanged.
```

File: pennylane/ops/qubit.py
Comment (antalszava) [inline pennylane/ops/qubit.py:1885]: The [rendering here seems to be a bit off](https://pennylane--1358.org.readthedocs.build/en/1358/code/api/pennylane.IsingYY.html?highlight=isingyy#pennylane.IsingYY), hopefully, the extra spaces will help:
Screenshot from 2021-06-14 13-18-35

The output after each change can be checked locally by running `make docs` or by having a look at the documentation continuous integration check at the bottom of the Pull Request. Clicking on the "Details" button will link to a changed version of the docs:

 Screenshot from 2021-06-14 13-21-02


There, it can be checked that this operation appears well e.g., by searching for `IsingYY`.
```suggestion
       where :math:`f` is an expectation value depending on :math:`YY(\phi)`.
```

File: pennylane/devices/tests/test_gates.py
Comment (antalszava) [reply pennylane/devices/tests/test_gates.py:199]: Got it! Which version of `black` was it? Our check is using `black-21.6b0`. It would be great to revert the formatting changes so that only the IsingYY addition is contained in this PR.

File: pennylane/ops/qubit.py
Comment (antalszava) [inline pennylane/ops/qubit.py:1603]: Just as a marker, it would be nice to have these changes reverted. That will help with looking through the real changes of the PR (even later down the line).

File: tests/devices/test_default_qubit.py
Comment (antalszava) [inline tests/devices/test_default_qubit.py:123]: Same note here too on reverting the formatting changes.

File: pennylane/devices/tests/test_gates.py
Comment (paniash) [reply pennylane/devices/tests/test_gates.py:199]: I'm on version `21.6b0`, so no version incompatibility.

File: pennylane/ops/qubit.py
Comment (antalszava) [inline pennylane/ops/qubit.py:1846]: Looks like the previous suggestion was incorrect. Maybe the best way is to have the gradient recipe definition be formatted just like the one for IsingXX (i.e., in two lines).

File: tests/ops/test_qubit_ops.py
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:81]: This change also seems to be formatting that could stay as was.

File: tests/ops/test_qubit_ops.py
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:2057]: This line seems to have an important change (to be kept).

File: tests/ops/test_qubit_ops.py
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:1440]: Most of the changes from here in this file could be reverted to shrink the scope  of the PR.

**Review by antalszava — COMMENTED** [review]
Hi @paniash, thanks for reverting formatting changes! There seem to be still a couple of occurrences in the tests, left some comments.

Also, it seems that the commit adding the changelog entry was reverted, please make sure that it's re-added such that this PR is documented in the changelog and you're added as a contributor! :slightly_smiling_face:

File: pennylane/ops/qubit.py
Comment (ixfoduap) [inline pennylane/ops/qubit.py:1827]:
```suggestion
    Ising YY coupling gate
```
For consistency with the other Ising XX and Ising ZZ operations

**Review by ixfoduap — APPROVED** [review]
Normally I would request more testing, but it's a simple operation and perhaps doesn't merit it

File: .github/CHANGELOG.md
Comment (antalszava) [inline .github/CHANGELOG.md:473]: Hi @paniash, could you move this addition to the top of the changelog where we have the details of version 0.17.0-dev? At the beginning of the week, we have entered a feature freeze stage for the PennyLane v0.16.0 release, which means that this new addition will come up in v0.17.0. :slightly_smiling_face: Also, don't forget to add yourself as one of the first contributors to the new release! :)

Once done, this PR is good to be merged.

Comment (paniash) [reply .github/CHANGELOG.md:473]: Awesome! Thanks. :-)

Comment (antalszava) [reply .github/CHANGELOG.md]: Thanks! Could you also add yourself to the new list of contributors? It will be starting on line 19. Preferably one new line before and after the names will be great.

Comment (paniash) [reply .github/CHANGELOG.md]: I had made a commit to the repo and added myself as a contributor at the time. Do I need to do it again specifically for 0.17.0-dev?

Comment (antalszava) [reply .github/CHANGELOG.md]: Yes, please. This contribution will be coming up in release 0.17.0. We have a group of contributors for each release separately, so in this case, you have contributed to both 0.16.0 and 0.17.0. We just need to mark the latter case too. :)

Comment (paniash) [reply .github/CHANGELOG.md]: Got it. Thanks!

**Pull request merged**
